### PR TITLE
Improve createToken interactor to create "app" tokens

### PR DESCRIPTION
**Pull request by esdrasjnr**
**Describe what this PR introduces:** This patch adds a parameter for user to choose the token type 
to be created.
**What kind of change does this PR introduce?** (check at least one)

- [ ] Bug fix
- [x] Feature
- [ ] Code style update
- [ ] Refactor
- [ ] Build-related changes
- [ ] Other, please describe:

**Does this PR introduce any breaking changes?** (check one)

- [ ] Yes
- [x] No

If yes, please describe the impact and migration path for existing applications:

**The PR fulfills these requirements:**

- [ ] Related issues are referenced in the PR (e.g. `fix #xxx[,#xxx]`, where "xxx" is the issue number)
- [ ] All tests are passing
- [ ] New/updated tests are included

**Testing environment:**

- Operating System/Platform: Ubuntu 18.04
- Go version: 1.13

If you have relevant logs, error output and etc, please attach them.

**Other information:**

### CODE REVIEW

File: pkg/user/delivery/http/user_proxy.go [hunk 1]
```diff
@@ -87,6 +102,42 @@ func (p *Proxy) CreateToken(user entities.User) (string, error) {
 	return tr.Token, nil
 }
 
+// CreateAppToken creates a valid token for the application
+func (p *Proxy) CreateAppToken(email, userToken string) (string, error) {
+	body, err := json.Marshal(&AppTokenRequest{Email: email, Type: 2, Duration: 3153600000})
```
Comment (netoax) [inline pkg/user/delivery/http/user_proxy.go]: What about creating a constant to store that duration value?

File: pkg/user/delivery/http/user_proxy.go [hunk 2]
```diff
@@ -87,6 +102,42 @@ func (p *Proxy) CreateToken(user entities.User) (string, error) {
...
+	if err != nil {
+		return "", fmt.Errorf("error creating request object: %w", err)
+	}
+	req.Header.Add("Authorization", userToken)
+	req.Header.Add("Content-Type", "application/json")
+
+	client := &http.Client{}
+	resp, err := client.Do(req)
+	if err != nil {
+		return "", fmt.Errorf("error executing POST request: %w", err)
```
Comment (netoax) [inline pkg/user/delivery/http/user_proxy.go]:
```suggestion
		return "", fmt.Errorf("error executing HTTP request: %w", err)
```

File: pkg/user/controllers/user.go [hunk 3]
```diff
@@ -17,6 +17,13 @@ type UserController struct {
 	createTokenInteractor *interactors.CreateToken
 }
 
+// CreateTokenParam represents the received parameters for CreateToken operation
+type CreateTokenParam struct {
```
Comment (netoax) [inline pkg/user/controllers/user.go]:
```suggestion
type CreateTokenRequest struct {
```

File: pkg/user/controllers/user.go [hunk 4]
```diff
@@ -83,23 +90,28 @@ func (uc *UserController) Create(w http.ResponseWriter, r *http.Request) {
...
 	if err != nil {
 		uc.logger.Error("failed to parse request body")
 		uc.writeResponse(w, http.StatusUnprocessableEntity, nil)
 		return
 	}
 
-	token, err := uc.createTokenInteractor.Execute(user)
+	tokenType := param.Type
+	if tokenType == "" {
+		tokenType = "user"
```
Comment (netoax) [inline pkg/user/controllers/user.go]:
```suggestion
	if req.Type == "" {
		req.Type = "user"
```

File: pkg/user/controllers/user.go [hunk 5]
```diff
@@ -83,23 +90,28 @@ func (uc *UserController) Create(w http.ResponseWriter, r *http.Request) {
 // @Router /tokens [post]
 // CreateToken handles the server request and calls CreateTokenInteractor
 func (uc *UserController) CreateToken(w http.ResponseWriter, r *http.Request) {
-	var user entities.User
-	err := json.NewDecoder(r.Body).Decode(&user)
+	var param CreateTokenParam
```
Comment (netoax) [inline pkg/user/controllers/user.go]:
```suggestion
	var req CreateTokenRequest
```

File: pkg/user/interactors/create_token.go [hunk 6]
```diff
@@ -20,8 +20,11 @@ func NewCreateToken(logger logging.Logger, userProxy http.UserProxy) *CreateToke
 
 // Execute receives the user entity filled with e-mail and password properties and try
 // to create a token on the user proxy service. If it succeed, the token is returned.
-func (ct *CreateToken) Execute(user entities.User) (token string, err error) {
+func (ct *CreateToken) Execute(user entities.User, tokenType string) (token string, err error) {
 	token, err = ct.userProxy.CreateUserToken(user)
+	if err == nil && tokenType == "app" {
```
Comment (netoax) [inline pkg/user/interactors/create_token.go]: If failed to create user token, the operation must be stopped, right? So, I think you can avoid validating if `err == nil` and move the `tokenType` checking to after the next if block.

```go
        if err != nil {
		ct.logger.Errorf("failed to create an user's token: %s", err.Error())
		return "", err
	}
        if tokenType == "app" {
                token, err = ct.userProxy.CreateAppToken(user.Email, token)
        }

        return token, nil


```

**Review by netoax — CHANGES_REQUESTED** [review]
Thanks!
I let some comments to help you improve this PR. Also, pay some attention to the commit messages.

- Commit tags such as `user:` must have a single word;
- Avoid to say application tokens have no expiration time, they actually have configurable duration.

File: pkg/user/controllers/user.go [hunk 7]
```diff
@@ -83,23 +90,28 @@ func (uc *UserController) Create(w http.ResponseWriter, r *http.Request) {
...
 		return
 	}
 
-	token, err := uc.createTokenInteractor.Execute(user)
+	tokenType := param.Type
+	if tokenType == "" {
+		tokenType = "user"
+	}
+
+	token, err := uc.createTokenInteractor.Execute(entities.User{Email: param.Email, Password: param.Password}, tokenType)
```
Comment (lcbm) [inline pkg/user/controllers/user.go]: Following @netoax's suggestion :point_down: 

```suggestion
	token, err := uc.createTokenInteractor.Execute(entities.User{Email: req.Email, Password: req.Password}, tokenType)
```

File: pkg/user/controllers/user.go [hunk 8]
```diff
@@ -83,23 +90,28 @@ func (uc *UserController) Create(w http.ResponseWriter, r *http.Request) {
...
 	if err != nil {
 		uc.logger.Errorf("failed to create user's token: %s", err)
 		der := &DetailedErrorResponse{err.Error()}
 		uc.writeResponse(w, mapErrorToStatusCode(err), der)
 		return
 	}
 
-	uc.logger.Infof("user's %s token created", user.Email)
+	uc.logger.Infof("user's %s token created", param.Email)
```
Comment (lcbm) [inline pkg/user/controllers/user.go]: Since we're updating this anyway :point_down: 

```suggestion
	uc.logger.Infof("token created for user %s", req.Email)
```

File: pkg/user/interactors/create_token.go (hunk 6 above)
Comment (lcbm) [reply pkg/user/interactors/create_token.go]: In addition to what @netoax said, I believe there is more room for improvement here :thinking:

If I understood correctly, we create a user token every time an app token is created. Is there a reason for this? Instead, I believe the user should create it's user token AND THEN, in a separate request, create the app token, with their user token. 

Otherwise, correct me if I'm wrong, the user would create a App Token and he would never know about the their new user token that was created, in the process. In addition, if they want to create more than one app token, they'll have to: 1) create user token, 2) create app token, 3) create new user token because they do not know the new token, 4) create app token, and this repeats itself for the amount of app tokens they want to to create.

**Review by lcbm — CHANGES_REQUESTED** [review]

**Comment by lcbm** [comment]
I accidentally forgot to comment on my review :stuck_out_tongue: 
I agree with the remarks made by @netoax and made a few other ones.

About the App Token duration... I don't believe it should be hard-coded! That's why we hassled for weeks, investigating ways to make it configurable... we are essentially creating the same barrier to our users (and limiting this awesome feature which is: having a configurable expiration time). 

Instead, I vouch for having a behavior similar to mainflux's: make it never expire by default, but allow the user to set a time if they wish to. It's a question of `If-else` for us and a world of possibilities for the user :smile:

File: pkg/user/delivery/http/user_proxy.go (hunk 1 above)
Comment (lcbm) [reply pkg/user/delivery/http/user_proxy.go]: Please see [this comment](https://github.com/CESARBR/knot-babeltower/pull/78#issuecomment-636193722) to understand the -1! :grimacing:

File: pkg/user/interactors/create_token.go (hunk 6 above)
Comment (netoax) [reply pkg/user/interactors/create_token.go]: You're right, @lcbm. The correct approach is to support this creation through another request that already comes with the user token. It's worth point this doesn't mean we need to implement two endpoints. What both of you think as a good design for that (cc. @lcbm, @esdrasjnr)?

Comment (lcbm) [reply pkg/user/interactors/create_token.go]: I'ts doable in the same endpoint but I still believe it should be two different operations. From what I understand, `type` parameter is already enough: 

1. check `type`
2. check paramters required for the appropriate `type`
3. return appropriate key or error 

We could also make the token duration configurable for api keys now, defaulting for never expiring if no `duration` params are passed.

Comment (lcbm) [reply pkg/user/interactors/create_token.go]: Hope [this](https://github.com/mainflux/mainflux/tree/master/authn) helps!

Comment (netoax) [reply pkg/user/interactors/create_token.go]: Agreed, @lcbm. 😄

File: pkg/user/delivery/http/user_proxy.go (hunk 2 above)
Comment (esdrasjnr) [reply pkg/user/delivery/http/user_proxy.go]: Done!

Comment (esdrasjnr) [reply pkg/user/delivery/http/user_proxy.go:122]: Done!

File: pkg/user/controllers/user.go (hunk 3 above)
Comment (esdrasjnr) [reply pkg/user/controllers/user.go]: Done!

File: pkg/user/controllers/user.go (hunk 4 above)
Comment (esdrasjnr) [reply pkg/user/controllers/user.go]: Done!

File: pkg/user/controllers/user.go (hunk 5 above)
Comment (esdrasjnr) [reply pkg/user/controllers/user.go]: Done!

File: pkg/user/controllers/user.go (hunk 7 above)
Comment (esdrasjnr) [reply pkg/user/controllers/user.go]: Done!

File: pkg/user/controllers/user.go (hunk 8 above)
Comment (esdrasjnr) [reply pkg/user/controllers/user.go]: Done!

File: pkg/user/delivery/http/user_proxy.go (hunk 1 above)
Comment (netoax) [reply pkg/user/delivery/http/user_proxy.go]: Agreed. Considering our time and needs, it's better to expose the feature in a similar way (but we don't necessarily need to expose as it is always).

File: pkg/user/delivery/http/user_proxy.go [hunk 9]
```diff
@@ -87,6 +102,42 @@ func (p *Proxy) CreateToken(user entities.User) (string, error) {
 	return tr.Token, nil
 }
 
+// CreateAppToken creates a valid token for the application
+func (p *Proxy) CreateAppToken(user entities.User) (string, error) {
+	body, err := json.Marshal(&AppTokenRequest{Email: user.Email, Type: 2, Duration: 3153600000})
```
Comment (netoax) [inline pkg/user/delivery/http/user_proxy.go]: As @lcbm suggested, duration should be configurable.

File: pkg/user/interactors/create_token.go (hunk 6 above)
Comment (esdrasjnr) [reply pkg/user/interactors/create_token.go]: Updated! Please, check if is there more improvements to be done :smile:

File: pkg/user/delivery/http/user_proxy.go (hunk 1 above)
Comment (esdrasjnr) [reply pkg/user/delivery/http/user_proxy.go]: I add a new commit to handle the duration time. Please check if it is ok! :smiley:

File: pkg/user/interactors/create_token_test.go [hunk 10]
```diff
@@ -27,6 +29,8 @@ var ctCases = []CreateTokenTestCase{
 		"token successfully created",
 		"user@user.com",
 		"123456789abcdef",
+		"user",
+		0,
```
Comment (ramonhpr) [inline pkg/user/interactors/create_token_test.go:44]: Please add tests changing theses properties.

File: pkg/user/delivery/http/user_proxy.go [hunk 11]
```diff
@@ -14,25 +14,40 @@ import (
...
+}
+
 // NewUserProxy creates a new Proxy instance
-func NewUserProxy(logger logging.Logger, hostname string, port uint16) *Proxy {
-	url := fmt.Sprintf("http://%s:%d", hostname, port)
-	logger.Debug("user proxy configured to " + url)
-	return &Proxy{url, logger}
+func NewUserProxy(logger logging.Logger, userHost string, userPort uint16, authHost string, authPort uint16) *Proxy {
+	userURL := fmt.Sprintf("http://%s:%d", userHost, userPort)
+	authURL := fmt.Sprintf("http://%s:%d", authHost, authPort)
```
Comment (ramonhpr) [inline pkg/user/delivery/http/user_proxy.go]: Since Authn is another service would not be correct to create an AuthnProxy?

File: pkg/user/controllers/user.go [hunk 12]
```diff
@@ -83,23 +92,31 @@ func (uc *UserController) Create(w http.ResponseWriter, r *http.Request) {
...
 		uc.writeResponse(w, http.StatusUnprocessableEntity, nil)
 		return
 	}
 
-	token, err := uc.createTokenInteractor.Execute(user)
+	if req.Type == "" {
+		req.Type = "user"
+	}
+	if req.Duration == 0 {
+		req.Duration = 31536000
```
Comment (ramonhpr) [inline pkg/user/controllers/user.go]: Can't use `time.Duration` or `time.Hour` ?

File: pkg/user/interactors/create_token.go [hunk 13]
```diff
@@ -20,12 +22,19 @@ func NewCreateToken(logger logging.Logger, userProxy http.UserProxy) *CreateToke
...
 // to create a token on the user proxy service. If it succeed, the token is returned.
-func (ct *CreateToken) Execute(user entities.User) (token string, err error) {
-	token, err = ct.userProxy.CreateUserToken(user)
+func (ct *CreateToken) Execute(user entities.User, tokenType string) (token string, err error) {
+	if tokenType == "user" {
+		token, err = ct.userProxy.CreateUserToken(user)
+	} else if tokenType == "app" {
+		token, err = ct.userProxy.CreateAppToken(user)
+	} else {
+		err = fmt.Errorf("invalid token type provided")
```
Comment (netoax) [inline pkg/user/interactors/create_token.go]:
```suggestion
		err = fmt.Errorf("only 'user' and 'app' token types are supported")
```

**Review by netoax — CHANGES_REQUESTED** [review]

File: pkg/user/delivery/http/user_proxy.go [hunk 14]
```diff
@@ -14,25 +14,40 @@ import (
...
+	CreateUserToken(user entities.User) (token string, err error)
+	CreateAppToken(user entities.User) (token string, err error)
 }
 
 // Proxy is responsible for implementing the user's proxy operations
 type Proxy struct {
-	url    string
-	logger logging.Logger
+	userURL string
+	authURL string
```
Comment (lcbm) [inline pkg/user/delivery/http/user_proxy.go]: Just a suggestion :point_down: 

```suggestion
	authnURL string
```
> If accepted, please remember to update all occurrences of this variable :smile:

File: pkg/user/controllers/user.go (hunk 12 above)
Comment (lcbm) [reply pkg/user/controllers/user.go]: Additionally, if this change is applied, how will this impact the user input? For example, as is now, I'm able to set the duration in seconds simply by passing some `int` value... if we update, will the input format change? If yes, how so?

**Review by lcbm — CHANGES_REQUESTED** [review]
Awesome update on the patch! I +1'd some comments @netoax and @ramonhpr made, im keeping an eye out for those. Also, a made a suggestion :smile:

File: pkg/user/delivery/http/user_proxy.go (hunk 9 above)
Comment (lcbm) [reply pkg/user/delivery/http/user_proxy.go]: This was updated in commit f5451ff83eb68c8f077b493abd02d60b0e75bf41

File: pkg/user/controllers/user.go (hunk 12 above)
Comment (ramonhpr) [reply pkg/user/controllers/user.go]: The Duration type is a int64. See: https://golang.org/pkg/time/#Duration
We can use ParseDuration too if we think that the user can pass like 1s or 2h. See: https://golang.org/pkg/time/#ParseDuration

File: pkg/user/interactors/create_token_test.go (hunk 10 above)
Comment (ramonhpr) [reply pkg/user/interactors/create_token_test.go:44]: Just completing this request if I wasn't clear. 
Add tests using empty string and `app` too.
Can add an property `functionInteractorMockcalled string` to check if the mock called the correct method

File: pkg/user/delivery/http/user_proxy.go (hunk 14 above)
Comment (esdrasjnr) [reply pkg/user/delivery/http/user_proxy.go]: OK! I gonna add a new commit changing both `authURL` to `authnURL` and `userURL` to `usersURL`. What you think?

File: pkg/user/delivery/http/user_proxy.go (hunk 11 above)
Comment (esdrasjnr) [reply pkg/user/delivery/http/user_proxy.go]: It's right! I gonna add a new file `authn_proxy.go` in the same directory

File: pkg/user/controllers/user.go (hunk 12 above)
Comment (lcbm) [reply pkg/user/controllers/user.go]: Awesome :heart:

File: pkg/user/delivery/http/user_proxy.go (hunk 14 above)
Comment (lcbm) [reply pkg/user/delivery/http/user_proxy.go]: No problem! Looks good to me :smile:

File: pkg/user/delivery/http/user_proxy.go (hunk 11 above)
Comment (esdrasjnr) [reply pkg/user/delivery/http/user_proxy.go]: Done!

File: pkg/user/delivery/http/user_proxy.go (hunk 14 above)
Comment (esdrasjnr) [reply pkg/user/delivery/http/user_proxy.go]: Since I create a new `authn_proxy.go`, I removed that prefix and let `URL`

File: pkg/user/interactors/create_token.go (hunk 13 above)
Comment (esdrasjnr) [reply pkg/user/interactors/create_token.go]: Done!

File: pkg/user/interactors/create_token_test.go (hunk 10 above)
Comment (esdrasjnr) [reply pkg/user/interactors/create_token_test.go:44]: Done!

File: pkg/user/controllers/user.go (hunk 12 above)
Comment (esdrasjnr) [reply pkg/user/controllers/user.go]: I end up deciding to not use it for some reasons:
- the `Duration` type do not provide a more simple way to handle long slices of time such days and years
- the use of `Duration` type will make the code a little bit more complex, adding error checking while creating a `Duration` variable
- it also will need to add two additional operations: convert the seconds received from client request to `time.Duration`, and finally convert the `Duration` variable to seconds, to be able to send a mainflux authn proxy request

If you still think it's better to use it instead of raw integers, please comment here! :smile: :+1:

File: pkg/user/delivery/http/user_proxy.go (hunk 9 above)
Comment (esdrasjnr) [reply pkg/user/delivery/http/user_proxy.go]: Updating commit hash: [103873](https://github.com/CESARBR/knot-babeltower/pull/78/commits/103873ccf530ee53bfc53ba95a5454245b34ba17)

File: pkg/user/delivery/http/users_proxy.go [hunk 15]
```diff
@@ -11,32 +11,32 @@ import (
 	"github.com/CESARBR/knot-babeltower/pkg/user/entities"
 )
 
-// UserProxy represents the interface to the user's proxy operations
-type UserProxy interface {
+// UsersProxy represents the interface to the user's proxy operations
+type UsersProxy interface {
 	Create(user entities.User) (err error)
-	CreateToken(user entities.User) (string, error)
+	CreateUserToken(user entities.User) (token string, err error)
```
Comment (netoax) [inline pkg/user/delivery/http/users_proxy.go]: Considering `UsersProxy` doesn't have `CreateAppToken` anymore, could we rename this method to just`CreateToken`? We know that's a user's token creation by the context.

**Review by netoax — CHANGES_REQUESTED** [review]
_103873ccf530ee53bfc53ba95a5454245b34ba17_: [...], this patch introduces **a new parameter to control the duration** time of the token, in seconds.
_f12f16c8b6ba321c7bf0fd28f6bbce5cfaee6845_: [...] In addition, add a CreateAppToken method **which** creates **a API** key with a long expiration time (precisely one year) to be used for applications.

File: pkg/user/delivery/http/users_proxy.go (hunk 15 above)
Comment (esdrasjnr) [reply pkg/user/delivery/http/users_proxy.go]: Yes! Gonna change it! :+1: :smiley:

Comment (netoax) [reply pkg/user/delivery/http/users_proxy.go]: Okay, let me know when you update this PR.

Comment (esdrasjnr) [reply pkg/user/delivery/http/users_proxy.go]: Done!

File: pkg/user/controllers/user.go (hunk 12 above)
Comment (ramonhpr) [reply pkg/user/controllers/user.go]: Seems good points to me. Resolving this comment

**Review by netoax — APPROVED** [review]
LGTM!
